### ISSUE TITLE
[BUG] MemoryError when trying to back up to external HDD (sudo needed) on server

### ISSUE THREAD

**Issue by morkr** [rdiff-backup/rdiff-backup#413]
## Bug summary

I am not sure if it is a bug or if I am just not experienced enough to see my mistake.
I try to backup a directory from my desktop to my server's encrypted external HDD via a bash script and get "MemoryError".

## Version, Python, Operating System

Both Desktop and Server are running:
rdiff-backup version 1.2.8, Python 2.7.17, Ubuntu 18.04.4 LTS

Call `rdiff-backup -v9` and paste the output between the backquotes below, repeat for each environment impacted:

Desktop
```
Thu Jul  2 17:41:59 2020  Using rdiff-backup version 1.2.8
Thu Jul  2 17:41:59 2020  Fatal Error: No arguments given
See the rdiff-backup manual page for more information.
```

Server
```
Thu Jul  2 15:43:10 2020  Using rdiff-backup version 1.2.8
Thu Jul  2 15:43:10 2020  Fatal Error: No arguments given
See the rdiff-backup manual page for more information.
```

## rdiff-backup call

How did you call rdiff-backup, with which parameter:

```
rdiff-backup --terminal-verbosity 5 --remote-schema 'ssh -C -tt %s sudo rdiff-backup --server' /home/user/testdir/ servername::/media/backup-drives/remote-backup/user-testPad/
```
I ran it inside a bash script, therefore the '-tt' option. I need root-privileges to write to the HDD. Would be nice if I didn't have to change that.

The first part of my script opens and mounts the HDD at the server via ssh (also using sudo and '-t' option). No problems there. Just for reference:
```
ssh -t servername 'sudo ~/scripts/remote-backup_serverside_1.sh'
```

## What happened and what did you expect?

The backup fails with this output:
```
Using rdiff-backup version 1.2.8
Executing ssh -C -tt servername sudo rdiff-backup --server
Exception '' raised of class '<type 'exceptions.MemoryError'>':
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/Main.py", line 304, in error_check_Main
    try: Main(arglist)
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/Main.py", line 321, in Main
    rps = map(SetConnections.cmdpair2rp, cmdpairs)
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/SetConnections.py", line 76, in cmdpair2rp
    if cmd: conn = init_connection(cmd)
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/SetConnections.py", line 152, in init_connection
    check_connection_version(conn, remote_cmd)
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/SetConnections.py", line 160, in check_connection_version
    try: remote_version = conn.Globals.get('version')
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/connection.py", line 450, in __call__
    return apply(self.connection.reval, (self.name,) + args)
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/connection.py", line 368, in reval
    result = self.get_response(req_num)
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/connection.py", line 315, in get_response
    try: req_num, object = self._get()
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/connection.py", line 240, in _get
    data = self._read(length)
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/connection.py", line 210, in _read
    try: return self.inpipe.read(length)

Traceback (most recent call last):
  File "/usr/bin/rdiff-backup", line 30, in <module>
    rdiff_backup.Main.error_check_Main(sys.argv[1:])
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/Main.py", line 304, in error_check_Main
    try: Main(arglist)
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/Main.py", line 321, in Main
    rps = map(SetConnections.cmdpair2rp, cmdpairs)
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/SetConnections.py", line 76, in cmdpair2rp
    if cmd: conn = init_connection(cmd)
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/SetConnections.py", line 152, in init_connection
    check_connection_version(conn, remote_cmd)
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/SetConnections.py", line 160, in check_connection_version
    try: remote_version = conn.Globals.get('version')
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/connection.py", line 450, in __call__
    return apply(self.connection.reval, (self.name,) + args)
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/connection.py", line 368, in reval
    result = self.get_response(req_num)
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/connection.py", line 315, in get_response
    try: req_num, object = self._get()
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/connection.py", line 240, in _get
    data = self._read(length)
  File "/usr/lib/python2.7/dist-packages/rdiff_backup/connection.py", line 210, in _read
    try: return self.inpipe.read(length)
MemoryError
```

I expected the backup to complete as it usually does locally and also when I execute sth like the following (writing to /home/user/ instead of /media/backup-drives/ and consequently not using 'sudo'):
```
rdiff-backup --terminal-verbosity 5 /home/user/testdir/ servername::/home/user/remote-backup/user-testPad/
```

I would be happy if someone could point me into the right direction or help me otherwise.

Have a great day!

## More information

Since nothing is backed up, there is also no log file in rdiff-backup-data (which does not exist itself), if I understand correctly.

**Comment by ikus060** [comment]
Could check if it's a duplicate of #355, #115.
See https://github.com/rdiff-backup/rdiff-backup/issues?q=MemoryError

**Comment by ericzolf** [comment]
Remove the `-tt`, it's the culprit for the MemoryError, and make sure with `sudo -n` that it doesn't need a password (if it does, you'll get a `sudo: a password is required` message). Please close the issue if that was the problem.

**Comment by morkr** [comment]
> Could check if it's a duplicate of #355, #115.
> See https://github.com/rdiff-backup/rdiff-backup/issues?q=MemoryError

Thanks for the suggestion, The two seemed related at first, but i think the causes are different.


> Remove the `-tt`, it's the culprit for the MemoryError, and make sure with `sudo -n` that it doesn't need a password (if it does, you'll get a `sudo: a password is required` message). Please close the issue if that was the problem.

Thank's for the quick support!
I tried that and got the (expected) error:

```
sudo: a password is required
Fatal Error: Truncated header string (problem probably originated remotely)
Couldn't start up the remote connection by executing
    ssh -C servername sudo -n rdiff-backup --server
```

Because I need the password.

So do I understand correctly: rdiff-backup can not handle the ssh option '-t' or '-tt' for piping/channeling/tunneling an interactive prompt from the server?
If so, I guess I have to avoid the need for a sudo-password to write on the HDD. Or find a different tool for the job.

Anyway: Great and easy to use software, this rdiff-backup. Was my first extensivly used CLI tool and I found the documentation easy to understand and helpful.

**Comment by ericzolf** [comment]
I've never needed `-t` or `-tt` but I think that you're misinterpreting, it sounds like it's offering a terminal emulation across the net, which is IMHO nothing sudo ever needs, rather utilities with a text UI they draw. The issue is rather that rdiff-backup is talking with the remote rdiff-backup server through stdout/stdin and if something, in this case sudo, goes in the way, e.g. by asking for a password, and tries to take away either stdin or stdout, it goes mixed up and it goes wrong.

This is what leads to the MemoryError: the communication from sudo is interpreted as a string length by rdiff-backup, which happens to be a huge length, which goes above the memory available. We've add already this discussion and the problem is that the only way to handle this properly would be to introduce checksums everywhere, with serious complication and performance implications.

Long story short: either make sure you can use sudo without password, and you should be fine, or connect directly as root with `root@servername::/your/path`, or even make sure that the target folder is writable by your normal user, which depending on your file system type should be IMHO the safest approach.

**Comment by morkr** [comment]
Have thanks for the thorough explanation and forgive my long absence.

I think I understood the problem now and will find a different solution (probably changing permissions for the write-directory).

Issue solved and closed.
